src/MongoDB.Driver/MongoCollectionImpl.cs

```diff
@@ -1017,6 +1019,22 @@ private TResult ExecuteWriteOperation<TResult>(IClientSessionHandle session, IWr
             }
         }
 
+        private ReadPreference GetReadPreference(IClientSession session)
```

Comment:
The logic is already at the ExecuteReadOperation level.

This is a helper method that is called from ExecuteReadOperation and ExecuteReadOperationAsync. It is factored into a helper method to avoid code duplication.

Would renaming the method help? GetReadPreferenceForReadOperation?

Reply:
I meant IOperationExecutor.ExecuteReadOperation/ExecuteReadOperationAsync

Reply:
Oh no...!

IOperationExecutor is a shim that was inserted for testing purposes.

It shouldn't have any logic.

Reply:
Ah, ok.  Java driver is quite different.  There's loads of logic there.

Reply:
Actually, one more thing occurred to me.  MongoDatabase.RunCommand can also execute in a transaction.  Don't we have to duplicate this logic there?

Reply:
Here's some sample code illustrating the odd behavior mandated by the spec:

```csharp
using (var session = client.StartSession())
{
    var transactionOptions = new TransactionOptions(readPreference: ReadPreference.Secondary);
    session.StartTransaction(transactionOptions);

    var command = new BsonDocument("ping", 1);
    var result = database.RunCommand<BsonDocument>(command, ReadPreference.Primary); // sent to Secondary, REALLY???
}
```

Reply:
Yeah, it's treated no different than:

```csharp
using (var session = client.StartSession())
{
    var transactionOptions = new TransactionOptions(readPreference: ReadPreference.Secondary);
    session.StartTransaction(transactionOptions);
    var result = database.GetCollection("test").WithReadPreference(ReadPreference.Secondary)
                       .count();
}
```

RunCommand is a special case already since by default it ignores the read preference set on the MongoDatabase.

Anyway, sounds like it's working to spec now so I think we're ok for this PR.  Correct?

Reply:
I don't think we're to spec (but the spec is weird and probably wrong).

Currently, if an explicit readPreference is passed to ReadCommand we are going to use that. The transaction's readPreference (if any) is not used.

If we need to make changes we'll open a new PR. This one is closed and the commit is pushed to master.

Reply:
Sounds good.  There's a ticket to add tests for RunCommand so it will be obvious if we're out of spec. https://jira.mongodb.org/browse/SPEC-1084